src/utils/token.ts

```diff
@@ -15,3 +31,184 @@ export function isOneToken(tokenName: TokenName | string, chainId: ChainId): boo
+        Number(await tokenContract.balanceOf(getToken(TokenName.ICHI_V2, chainId).address)) / 10 ** 9;
+      const ichiV2GSRBalance = Number(await ichiV2Contract.balanceOf(getAddress(AddressName.GSR, chainId))) / 10 ** 18;
+
+      circulating =
+        totalTokens +
+        ichiV2TotalSupply -
+        v1Balance -
+        v2Balance -
+        v3Balance -
+        ichiInV2Balance -
+        communityGnosisBalance -
+        ichiV2GnosisBalance -
+        ichiAllyBalance -
+        ichiV2GSRBalance;
+
+      // console.log(circulating);
+    }
+
+    if (tokenName == TokenName.ICHI_V2) {
+      const ichiContract = getErc20Contract(getToken(TokenName.ICHI, chainId).address, provider);
+      let ichiTotalSupply = Number(await ichiContract.totalSupply()) / 10 ** 9;
+
+      const v1Balance = Number(await ichiContract.balanceOf(getAddress(AddressName.FARMING_V1, chainId))) / 10 ** 9;
+      const v2Balance = Number(await ichiContract.balanceOf(getAddress(AddressName.FARMING_V2, chainId))) / 10 ** 9;
+      const v3Balance = Number(await tokenContract.balanceOf(getAddress(AddressName.FARMING_V3, chainId))) / 10 ** 18;
+      const communityGnosisBalance =
+        Number(await ichiContract.balanceOf(getAddress(AddressName.ICHI_COMMUNITY_GNOSIS, chainId))) / 10 ** 9;
+      const ichiV2GnosisBalance =
+        Number(await tokenContract.balanceOf(getAddress(AddressName.ICHI_V2_GNOSIS, chainId))) / 10 ** 18;
+      const ichiAllyBalance = Number(await tokenContract.balanceOf(getAddress(AddressName.ALLY, chainId))) / 10 ** 18;
+      const ichiInV2Balance =
+        Number(await ichiContract.balanceOf(getToken(TokenName.ICHI_V2, chainId).address)) / 10 ** 9;
+      const ichiGSRBalance = Number(await tokenContract.balanceOf(getAddress(AddressName.GSR, chainId))) / 10 ** 18;
+
+      circulating =
+        totalTokens +
+        ichiTotalSupply -
+        v1Balance -
+        v2Balance -
+        v3Balance -
+        ichiInV2Balance -
+        communityGnosisBalance -
+        ichiV2GnosisBalance -
+        ichiAllyBalance -
+        ichiGSRBalance;
+    }
+
+    if ([TokenName.ONE_BTC, TokenName.ONE_UNI, TokenName.ALLY].includes(tokenName)) {
+      // special case - get price from oneToken/ICHI vault's pool
+      const ichiV2Token = getToken(TokenName.ICHI_V2, chainId);
+
+      let ichiPrice: number;
+
+      // If tokenPrices has been passed in defer to that if possible
+      if (opts.tokenPrices && ichiV2Token.address.toLowerCase() in opts.tokenPrices) {
+        ichiPrice = opts.tokenPrices[ichiV2Token.address.toLowerCase()].usd;
+      } else {
+        let tokenPrices = await lookUpTokenPrices([ichiV2Token.address.toLowerCase()]);
+        if (!tokenPrices || !(ichiV2Token.address.toLowerCase() in tokenPrices)) {
+          throw new Error(`Could not lookup token prices for ${ichiV2Token.symbol}, possibly flooding CoinGecko`);
+        }
+
+        ichiPrice = tokenPrices[ichiV2Token.address.toLowerCase()].usd;
```

Comment:
@vastreck Here is an example of the new logic, prefer the opts.tokenPrices, if that doesn't resolve, only then dynamically lookup the prices